File: libindy/tests/utils/pairwise.rs
extern crate futures;

use indy::IndyError;
use indy::pairwise;
use self::futures::Future;

use indy::WalletHandle;

pub fn pairwise_exists(wallet_handle: WalletHandle, their_did: &str) -> Result<bool, IndyError> {
    pairwise::is_pairwise_exists(wallet_handle, their_did).wait()
}

pub fn create_pairwise(wallet_handle: WalletHandle, their_did: &str, my_did: &str, metadata: Option<&str>) -> Result<(), IndyError> {
    pairwise::create_pairwise(wallet_handle, their_did, my_did, metadata).wait()
}

pub fn list_pairwise(wallet_handle: WalletHandle) -> Result<String, IndyError> {
    pairwise::list_pairwise(wallet_handle).wait()
}

pub fn get_pairwise(wallet_handle: WalletHandle, their_did: &str) -> Result<String, IndyError> {
    pairwise::get_pairwise(wallet_handle, their_did).wait()
}

pub fn set_pairwise_metadata(wallet_handle: WalletHandle, their_did: &str, metadata: Option<&str>) -> Result<(), IndyError> {
    pairwise::set_pairwise_metadata(wallet_handle, their_did, metadata).wait()
}

File: wrappers/nodejs/test/pairwise.js
const test = require('ava')
const indy = require('../')
const cuid = require('cuid')
const initTestPool = require('./helpers/initTestPool')

test('pairwise', async function (t) {
  const pool = await initTestPool()
  const walletConfig = { id: 'wallet-' + cuid() }
  const walletCredentials = { key: 'key' }
  await indy.createWallet(walletConfig, walletCredentials)
  const wh = await indy.openWallet(walletConfig, walletCredentials)

  const [theirDid, theirVerkey] = await indy.createAndStoreMyDid(wh, {})
  const [myDid] = await indy.createAndStoreMyDid(wh, {})
  await indy.storeTheirDid(wh, { did: theirDid, verkey: theirVerkey })

  t.deepEqual(await indy.listPairwise(wh), [])

  t.false(await indy.isPairwiseExists(wh, theirDid))

  await indy.createPairwise(wh, theirDid, myDid, 'wat')

  t.true(await indy.isPairwiseExists(wh, theirDid))

  t.deepEqual(await indy.listPairwise(wh), [
    { my_did: myDid, their_did: theirDid, metadata: 'wat' }
  ])

  t.deepEqual(await indy.getPairwise(wh, theirDid), {
    my_did: myDid,
    metadata: 'wat'
  })

  await indy.setPairwiseMetadata(wh, theirDid, 'hello new metadata')

  t.deepEqual(await indy.getPairwise(wh, theirDid), {
    my_did: myDid,
    metadata: 'hello new metadata'
  })

  await indy.closeWallet(wh)
  await indy.deleteWallet(walletConfig, walletCredentials)
  pool.cleanup()
})
